pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,310 @@
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	if err := s.maybeFlushOOOSpans(ctx, newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	var isContiguous, isSame bool
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		isSame = lastFile.Span.Equal(newSpan)
+		isContiguous = isContiguousSpan(lastFile.Span, newSpan)
+		if isSame {
+			return nil
+		} else if isContiguous &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
+			s.stats.spanGrows++
+			lastFile.Span.EndKey = newSpan.EndKey
+			return nil
+		}
+	}
+
+	// If we are resetting to a new span that is not contiguous with the
+	// last span, we need to ensure that we are not mid-row.
+	if lastFile != nil && !isContiguous && s.midRow {
+		if s.midRow {
+			return errors.AssertionFailedf(
+				"backup closed file ending mid-key in %v", lastFile.Span.EndKey,
+			)
+		}
+	}
+	s.flushedFiles = append(
+		s.flushedFiles,
+		backuppb.BackupManifest_File{
+			Span: newSpan,
+			Path: s.outName,
+		},
+	)
+	return nil
+}
+
+func (s *SSTSinkKeyWriter) Flush(ctx context.Context) error {
+	if len(s.flushedFiles) > 0 {
+		lastFile := s.flushedFiles[len(s.flushedFiles)-1]
+		// If WriteKey was used, it is possible that the last written key was
+		// mid-row, but because no keys were written after, s.midRow was not updated.
+		// To ensure this, we update midRow using the end of the span that the last
+		// key belonged to.
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+	}
+	err := s.FileSSTSink.Flush(ctx)
+	s.prevKey = nil
+	return err
+}
+
+// TODO (kev-cao): Most of these `maybeFlushX` functions can be generalized to
+// other SSTSink implementations. When we get around to refactoring out the
+// other implementation, we should consider moving these functions into the base
+// FileSSTSink implementation.
+
+// maybeFlushOOOSpans checks if the new span's startKey precedes the last spans'
+// endKey or if the two spans do not share the same prefix based on the sink's
+// ElideMode. If either of these conditions are met, the current backup file is
+// flushed and a new one is created.
+func (s *SSTSinkKeyWriter) maybeFlushOOOSpans(ctx context.Context, newSpan roachpb.Span) error {
+	if len(s.flushedFiles) == 0 {
+		return nil
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if newSpan.Equal(lastFile.Span) {
```

Comment:
why would this ever happen?

Reply:
Updated now with the assumption that having the same span should result in a flush.

Reply:
To be clear, i don't think the user will ever pass a repeat span to the sink. We could throw an error.